openhands/runtime/utils/edit.py

```diff
@@ -0,0 +1,354 @@
+NOTE:
+- The output file should be COMPLETE and CORRECTLY INDENTED. Do not omit any lines, and do not change any lines that are not part of the changes.
+- You should output the new version of the file by wrapping the new version of the file content in a ``` block.
+- If there's no explicit comment to remove the existing code, we should keep them and append the new code to the end of the file.
+- If there's placeholder comments like `# no changes before` or `# no changes here`, we should replace these comments with the original code near the placeholder comments.
+"""
+
+USER_MSG = """
+HERE IS THE OLD VERSION OF THE FILE:
+```
+{old_contents}
+```
+
+HERE IS THE DRAFT OF THE NEW VERSION OF THE FILE:
+```
+{draft_changes}
+```
+
+GIVE ME THE NEW VERSION OF THE FILE.
+IMPORTANT:
+- There should be NO placeholder comments like `# no changes before` or `# no changes here`. They should be replaced with the original code near the placeholder comments.
+- The output file should be COMPLETE and CORRECTLY INDENTED. Do not omit any lines, and do not change any lines that are not part of the changes.
+""".strip()
+
+
+def _extract_code(string):
+    pattern = r'```(?:\w*\n)?(.*?)```'
+    matches = re.findall(pattern, string, re.DOTALL)
+    if not matches:
+        return None
+    return matches[0]
+
+
+def get_new_file_contents(
+    llm: LLM, old_contents: str, draft_changes: str, num_retries: int = 3
+) -> str | None:
+    while num_retries > 0:
+        messages = [
+            {'role': 'system', 'content': SYS_MSG},
+            {
+                'role': 'user',
+                'content': USER_MSG.format(
+                    old_contents=old_contents, draft_changes=draft_changes
+                ),
+            },
+        ]
+        resp = llm.completion(messages=messages)
+        new_contents = _extract_code(resp['choices'][0]['message']['content'])
+        if new_contents is not None:
+            return new_contents
+        num_retries -= 1
+    return None
+
+
+class FileEditRuntimeInterface(ABC):
+    config: AppConfig
+
+    @abstractmethod
+    def read(self, action: FileReadAction) -> Observation:
+        pass
+
+    @abstractmethod
+    def write(self, action: FileWriteAction) -> Observation:
+        pass
+
+
+class FileEditRuntimeMixin(FileEditRuntimeInterface):
+    # Most LLMs have output token limit of 4k tokens.
+    # This restricts the number of lines we can edit to avoid exceeding the token limit.
+    MAX_LINES_TO_EDIT = 300
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        llm_config = self.config.get_llm_config()
+
+        if llm_config.draft_editor is None:
+            logger.warning(
+                f'Draft editor LLM is not set. Fallback to use the LLM (model: {llm_config.model}) for the main agent. '
+                'Note that this may incur high costs for editing actions.'
+            )
+            llm_config.draft_editor = copy.deepcopy(llm_config)
```

Comment:
As we discussed face-to-face, the cost may not be so high. Do we just want to remove this warning?

Reply:
removed!